prql-compiler/tests/integration/connection.rs

```diff
@@ -8,339 +8,366 @@ use chrono::{DateTime, Utc};
-                        "null".to_string()
-                    } else {
-                        format!("\"{}\"", s.replace('"', "\"\""))
-                    })
-                    .join(",")
-            );
-            self.run_query(q.as_str(), runtime).unwrap();
-        }
-    }
-
-    fn get_dialect(&self) -> Dialect {
-        Dialect::SQLite
+            Dialect::SQLite => {
+                let conn = rusqlite::Connection::open_in_memory()?;
+                Protocol::Sqlite(conn)
+            }
+            Dialect::MySql => {
+                let conn = mysql::Pool::new("mysql://root:root@localhost:3306/dummy")?;
+                Protocol::MySql(conn)
+            }
+            Dialect::Postgres => {
+                let conn = postgres::Client::connect(
+                    "host=localhost user=root password=root dbname=dummy",
+                    postgres::NoTls,
+                )?;
+                Protocol::Postgres(conn)
+            }
+            Dialect::MsSql => {
+                let mut config = Config::new();
+                config.host("127.0.0.1");
+                config.port(1433);
+                config.trust_cert();
+                config.authentication(AuthMethod::sql_server("sa", "Wordpass123##"));
+                use tokio_util::compat::TokioAsyncWriteCompatExt;
+                let conn = RUNTIME.block_on(async {
+                    let tcp = TcpStream::connect(config.get_addr()).await?;
+                    tcp.set_nodelay(true).unwrap();
+                    tiberius::Client::connect(config, tcp.compat_write()).await
+                })?;
+                Protocol::SqlServer(conn)
+            }
+            _ => unimplemented!("integration is not implemented for {:?}", dialect),
+        };
+        Ok(DBConnection { dialect, protocol })
     }
-}
-
-impl DBConnection for postgres::Client {
-    fn run_query(&mut self, sql: &str, _runtime: &Runtime) -> Result<Vec<Row>> {
-        let rows = self.query(sql, &[])?;
-        let mut vec = vec![];
-        for row in rows.into_iter() {
-            let mut columns = vec![];
-            for i in 0..row.len() {
-                let col = &(*row.columns())[i];
-                let value = match col.type_() {
-                    &Type::BOOL => (row.get::<usize, bool>(i)).to_string(),
-                    &Type::INT4 => match row.try_get::<usize, i32>(i) {
-                        Ok(v) => v.to_string(),
-                        Err(_) => "".to_string(),
-                    },
-                    &Type::INT8 => match row.try_get::<usize, i64>(i) {
-                        Ok(v) => v.to_string(),
-                        Err(_) => "".to_string(),
-                    },
-                    &Type::TEXT | &Type::VARCHAR | &Type::JSON | &Type::JSONB => {
-                        match row.try_get::<usize, String>(i) {
+    pub fn run_query(&mut self, sql: &str, runtime: &Runtime) -> Result<Vec<Row>> {
+        match self.protocol {
```

Comment:
My initial reaction is that these huge match blocks for every method of `DBConnection` are an indicator that this should be a trait implemented on each of instances, like the previous approach.

I agree it would be good to separate the protocol from the dialect eventually (though it's also fine to wait until we have the DBs :) ). But can we do that by returning a `DBConnection` here? https://github.com/PRQL/prql/pull/2804/files#diff-61c6edc4c83a1ee02e8c718cd4c3fe7957d91b357a60b89daed30876c1544b3bR44. We'd have something like `Dialect::DataBend => mysql::Pool::new("databend://root:root@localhost:3307/dummy").unwrap()`

Indeed — isn't `DBConnection` the `Protocol` now?

Reply:
Thanks for checking this.

Since only `run_query` can be shared between some dialects (e.g. MySQL and ClickHouse), something like `import_csv` may be separated by dialect, I believe it is necessary here to return a structure with dialect information, not a protocol.

Reply:
But if `import_csv` doesn't work for that dialect — e.g. the querying is the same protocol but not the loading — then we can make a new Struct and implement different code on it, forwarding to `MySql::pool` where necessary.

I don't want to be discouraging, and it looks like you've done a lot of rust which is great. But I still think that the big match on each method is some evidence that this is not the best data model. And I'm not sure what we're getting for the extra layer of indirection? I'm very open to being wrong if @aljazerzen has a perspective?

I don't think it's _bad_ tbc, I don't quite agree with the implementation, that's all, I hope that's reasonable...

I do think that the current `.get_dialect` is bad, let me take a few minutes to try and remove that. (which will cause a small merge conflict but I hope is OK)

Reply:
I agree that this does not seem to be the best thing to do.
It is simply the only way I could have done it.

If you could rewrite this for me that would be great. (Of course I had hoped this would not be too much of a burden on you, but I am not up to the task)

Reply:
> If you could rewrite this for me that would be great. (Of course I had hoped this would not be too much of a burden on you, but I am not up to the task)

I really didn't mean to be discouraging! It's overall been very impressive how far your rust has come...

It's very possible I'm wrong here — generally each person has something right and something wrong — that's why I tagged others.

I had a go in #2806, let me know your thoughts.

Reply:
Thank you very much. I don't think I am being discouraged. Don't worry about it.

> I had a go in #2806, let me know your thoughts.

Awesome, that makes more sense!

Reply:
I looked at this and couldn't figure out how to implement it without using match.
I've created an issue #2813 and will leave it to someone else.